kernel/src/kernel.rs

```diff
@@ -445,18 +446,30 @@ impl Kernel {
     ///
     /// Most of the behavior of this loop is controlled by the `Scheduler`
     /// implementation in use.
-    pub fn kernel_loop<KR: KernelResources<C>, C: Chip, const NUM_PROCS: u8>(
+    pub fn kernel_loop<KR: KernelResources<C>, C: Chip + ChipAtomic, const NUM_PROCS: u8>(
         &self,
         resources: &KR,
         chip: &C,
         ipc: Option<&ipc::IPC<NUM_PROCS>>,
         capability: &dyn capabilities::MainLoopCapability,
+        do_closure: bool,
+        closure: Option<&dyn Fn()>,
     ) -> ! {
         resources.watchdog().setup();
+
+        let mut hart_id: usize = 0;
+        unsafe {
+            core::arch::asm!("csrr {a}, mhartid", a = out(reg) hart_id);
+        }
+        let id = unsafe { DynThreadId::new(hart_id) };
```

Comment:
Not to review in too much detail at this stage, but I assume a future version would not require this assembly here.

Reply:
Yes, this exists purely for debugging purposes and should be removed later. All high-level kernel abstractions are currently not thread-aware, which is why a large portion of the portal mechanism is implemented in a board-specific way. This also indicates the kernel code is shared across multiple kernel instances which could be an issue if we want to enforce isolation (including timing) between kernel instances